### PR TITLE
Reform handling of quote characters in raw strings

### LINKED ISSUE #3959: Quote gets eaten as part of quadruple quote
## Description

When quadruple doublequotes appear in a file, it should be parsed as the start of a triple-quoted block starting with a single doublequote character. Instead, it seems that the quote simply disappears and does not appear in the generated file.

## To Reproduce

```purescript
module Test where

s :: String
s = """"should start with a quote"""
```

Generated output:

```js
// Generated by purs version 0.13.3
"use strict";
var s = "should start with a quote";
module.exports = {
    s: s
};

```

## Expected behavior

The generated string should start with a quote: `var s = "\"should start with a quote"`.

## Additional context

I believe this is a regression from 0.12, since I included this in code that worked fine before upgrading to 0.13.

## PureScript version

0.13.3 and 0.14 (master branch)

### PR DESCRIPTION
**Pull request by rhendric**
This corrects a number of regressions introduced in 0.13, and imposes
the uniform restriction that `"""` can't appear anywhere inside a raw
string literal. It's possible (in fact, slightly easier) to lex strings
such that an unlimited number of quotes are allowed at the end (and only
at the end) of a raw string literal, but the simplicity and clarity of a
uniform rule outweighs the implementation cost and the very minor loss
of expressiveness.

Closes #3959.

### CODE REVIEW

File: lib/purescript-cst/src/Data/Text/PureScript.hs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+-- |
+-- This module contains internal extensions to Data.Text.
+--
+module Data.Text.PureScript (spanUpTo) where
```
Comment (rhendric) [inline lib/purescript-cst/src/Data/Text/PureScript.hs:4]: So, this may be a questionable choice. It's possible to correctly implement `spanUpTo` without using internal and unsafe API from `text`, but only in what I imagine is a somewhat inefficient way—`span` does a Unicode scalar-to-code-point walk on the text, and then `compareLength` does the same, and then if the initial span is too large a `splitAt` does a third walk. I don't know if this makes a real world difference, is the thing. But implementing `spanUpTo` by just tweaking the internal implementation of `span`, as I did here, is very straightforward. I stuffed it in a private module here to limit the consequences of importing these parts of the `text` API elsewhere. As far as premature optimizations go, this one seems pretty benign. But then, they probably all do at first. I can easily rewrite it if this raises flags.

**Review by natefaubion — APPROVED** [review]

File: tests/purs/passing/BlockStringEdgeCases.purs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  assert' "ends with two quotes" ("""x""""" == "x\"\"")
+  assert' "starts and ends with two quotes" ("""""x""""" == "\"\"x\"\"")
+  assert' "mixture 1" ("""""x"y""z"""" == "\"\"x\"y\"\"z\"")
+  assert' "mixture 2" ("""x"y""z""" == "x\"y\"\"z")
+
+  -- These last tests are more about forbidding certain raw string literal
+  -- edge cases than about wanting to support mashing string literals against.
+  -- each other, which is techically legal but generally, if not universally,
+  -- a bad idea.
+  assert' "too many quotes 1" (Tuple """"""""" " == Tuple "\"\"" " ")
```
Comment (hdgarrood) [inline tests/purs/passing/BlockStringEdgeCases.purs:26]: It occurs to me that it might be worth trying to update the PureScript syntax highlighting because the syntax highlighting here is wrong, presumably? This is what it looks like for me:

Screenshot from 2020-12-16 03-04-06



The highlighting used on GitHub is defined in https://github.com/purescript-contrib/atom-language-purescript if I remember correctly. This is not a blocker for merging, just thought it's worth mentioning.

Comment (hdgarrood) [reply tests/purs/passing/BlockStringEdgeCases.purs:26]: Ah yes, I see what's going on here now; the highlighting is definitely wrong. It wouldn't compile if the parser understood the boundaries of these string literals the same way this highlighter did.

**Review by hdgarrood — APPROVED** [review]
